### ISSUE TITLE
How to process the file in `friend_and_hashtag_prediction_userids.zip `?

### ISSUE THREAD

**Issue by JyiHUO** [abenton/wgcca#4]
I did some exploration about `friend_and_hashtag_prediction_userids.zip `  and found that there were sever hundred row with each file:

![image](https://user-images.githubusercontent.com/16526817/43389725-0211e01e-941f-11e8-8a61-a9e8548618d3.png)

However, the training dataset in the paper have 102,328 Twitter users. So, is there something missing in this four files?

**Comment by abenton** [comment]
These files are described here [http://www.cs.jhu.edu/~mdredze/datasets/multiview_embeddings/](http://www.cs.jhu.edu/~mdredze/datasets/multiview_embeddings/)

> Each row in a file corresponds to a single hashtag or celebrity. The first field is the hashtag users posted or celebrity they follow. All following entries are the user IDs of everyone who engaged. The first 10 user IDs were used to compute the query embedding (rank all other user IDs by cosine similarity).

Number of rows in file corresponds to number of hashtags/celebrities.  Users who did not tweet a hashtag or follow a friend are not included in these files.

**Comment by JyiHUO** [comment]
Thanks for your reply. I do some data preprocessing to make sure that I did right to extract label for each row in the biggest dataset. And I show the code below:

![image](https://user-images.githubusercontent.com/16526817/43464098-c59ffc30-950c-11e8-9e1d-9d55feceab65.png)


![image](https://user-images.githubusercontent.com/16526817/43464117-d088bcb8-950c-11e8-83db-8f6a1ad87126.png)

The code above show that if the user follow the celebrities who are in the top 500 famous people. I should label this user '1' otherwise '0'

**So, did I do anything right or wrong?**

**Comment by abenton** [comment]
The way we evaluated embeddings in our paper is that we treated every row in `uidPerFriend_dev_all.txt` as a separate recommendation task.   Take for example the first row in `uidPerHashtag_dev_all.txt`:

>     #makinglightofasituation	1355121067	899826914	1727435834	1073335819	551173605	898735326	299725469	774405877	632582159	1657941475	979165920	1488580639	2156462635	619627618	701545538	229753245	145218701	160076675	1438028888	134354726	606807345	393401396	348553574	549095719	136345323	911510640	553106025	47119870	1472427212	132444758	712518024	844447854	445726276	1628810742	1145810940	149579296	1129499792	832165729

This is a hashtag, _#makinglightofasituation_, that a total of 39 other users tweeted.  Given a query of the average embedding of the first 10 user IDs, we want the other 29 users to be highly ranked by cosine similarity.  The way your code is written, you are constructing a different task: "Rank users and evaluate on whether they tweeted **any** of the 200 hashtags.", not "Rank users and evaluate on whether they tweeted this **particular** hashtag, #makinglightofasituation".  Labeling each example would look something like this (excuse bugs/typos, did not run this):

```python
labels_per_example = []  # label for every recommendation task for each user ID
for i in ids:
  labels = []
  for r in res:
    labels.append(1 if i in r else 0)
  labels_per_example.append(labels)
```

What we finally report in our paper is macro-averaged recall and precision over all 200 user engagement and 250 friend prediction tasks.  Does this make sense?

**Comment by JyiHUO** [comment]
Great. Thanks for your patient explanation. This indeed helps me a lot. After catching your point, I get the target data which seem like this:

![image](https://user-images.githubusercontent.com/16526817/43473423-a9bbefea-9522-11e8-9b8f-6e9532a8ff57.png)

**Comment by abenton** [comment]
Looks good
